drivers/gxe.c

```diff
@@ -0,0 +1,426 @@
+		upsdebugx(LOG_WARNING, "read: %s", ret ? strerror(errno) : "timeout");
+		return ret;
+	}
+
+	return ret;
+}
+
+int val_from_hex(char *buf, size_t dlen)
+{
+	char valbuf[16];
+
+	if (dlen > 15)
+		return 0;
+	
+	memcpy(valbuf, buf, dlen);
+	valbuf[dlen] = '\0';
+	return strtol(valbuf, NULL, 16);
+}
+
+void substr_from_asciihex(char *substr, char *dbuf, size_t dlen)
+{
+	int val;
+	size_t i;
+	for (i = 0; i < dlen; i += 2) {
+		val = val_from_hex(dbuf+i, 2);
+		if (val == 0x20) {
+			substr[i] = '\0';
+			return;
+		}
+		substr[i/2] = val;
+	}
+	substr[dlen/2] = '\0';
+	upsdebugx(LOG_DEBUG, "substr: %s", substr);
+}
+
+static int instcmd(const char *cmdname, const char *extra)
+{
+	upslogx(LOG_NOTICE, "instcmd: unknown command [%s]", cmdname);
+	return STAT_INSTCMD_UNKNOWN;
+}
+
+void upsdrv_initinfo(void)
+{
+	char databuf[64];
+	char recvbuf[64];
+	int retry, ret;
+	
+	for (retry = 0; retry < PROBE_RETRIES; retry++) {
+		ret = frame_send_and_recv(recvbuf, 64,
+					  GXE_GET_VENDOR_INFO,
+					  "21", devaddr, NULL, 0);
+		if (ret > 34)
+			break;
+	}
+
+	if (ret <= 0)
+		fatal_with_errno(EXIT_FAILURE, "gxe: failed reading response");
+	if (ret < 35)		/* Minimum Length for Name */
+		fatalx(EXIT_FAILURE, "gxe: not enough data");
+
+	/* UPS Name, 10 bytes */
+	substr_from_asciihex(databuf, recvbuf+13, 20);
+	dstate_setinfo("ups.mfr", "EpsonNetworkPower");
+	dstate_setinfo("ups.model", "%s", databuf);	
+
+	dstate_setinfo("ups.id", "%s", devaddr);
+	
+	dstate_addcmd("test.battery.start");
+	dstate_addcmd("test.battery.stop");
+	dstate_addcmd("load.off");
+	dstate_addcmd("load.on");
+
+	upsh.instcmd = instcmd;
+
+	poll_state = GXE_SYSPARM;
+}
+
+int validate_ret(char *buf, int ret, int minlen)
+{
+	if (ret < 0) {
+		upslog_with_errno(LOG_ERR, "Read failed");
+		return -1;
+	}
+	
+	if (ret < minlen) {
+		upslogx(LOG_ERR, "Short read from UPS");
+		return -1;
+	}
+
+	if ((ret = val_from_hex(buf+7, 2))) {
+		upslogx(LOG_ERR, "Command failed: %s", RTN_TO_STR(ret));
+		return -1;
+	}
+	
+	return ret;
+}
+
+void upsdrv_updateinfo_onoff(void)
+{
+	char onoff_buf[64];
+	int ret, dflag;
+	
+	ret = frame_send_and_recv(onoff_buf, 64,
+				  GXE_GET_ONOFF_DATA,
+				  "21", devaddr, NULL, 0);
+
+	if (validate_ret(onoff_buf, ret, 13+0x14) < 0) {
+		dstate_datastale();
+		return;
+	}
+
+	poll_state = GXE_ANALOG;
+
+	/* DATAFLAG */
+	dflag = val_from_hex(onoff_buf+13, 2);
+	if (dflag & DATAFLAG_ONOFF_MASK)
+		poll_state = GXE_ONOFF;
+	if (dflag & DATAFLAG_WARN_MASK)
+		poll_state = GXE_WARNING;
+
+	status_init();
+	/* Field 1, Power Supply */
+	status_set(val_from_hex(onoff_buf+15+2, 2) == 0x01 ? "OB" : "OL BYPASS");
+	status_commit();
+
+	/* Field 4, Battery Status */
+	switch(val_from_hex(onoff_buf+15+6, 2)) {
+	case 0xe0:
+		dstate_setinfo("battery.charger.status", "resting");
+		break;
+	case 0xe1:
+	case 0xe2:
+		dstate_setinfo("battery.charger.status", "charging");
+		break;
+	case 0xe3:
+		dstate_setinfo("battery.charger.status", "discharging");
+	}
```

Comment:
I think some compilers' static analyzers (e.g. `clang-16` or newer) would complain about lacks of `break` here, and of `default:` label in general. Also in other `switch/case` blocks here...